## Key dataframe operations
### selecting, sorting, adding and aggregating data in dataframes

#### import the Libraries

In [1]:
import pandas as pd
import numpy as np

#### Task 1: Read the nba.csv file and place it in a dataframe variable called `nba_df`.

In [2]:
nba_df = pd.read_csv('nba.csv')

#### Task 2: Show the first 5 records of your dataframe

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,6-6,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,NaN
3,R.J. Hunter,Boston Celtics,28.0,SG,22.0,6-5,185.0,Georgia State,1148640.0
4,Jonas Jerebko,Boston Celtics,8.0,PF,29.0,6-10,231.0,NaN,5000000.0


#### Task 3: Display a summary of the basic information about this DataFrame and its data.
- use the `info()` and `describe` functions
- What do these functions tell you about your dataframe?

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


,Number,Age,Weight,Salary
count,457.000000,457.000000,457.000000,4.460000e+02
mean,17.678337,26.938731,221.522976,4.842684e+06
std,15.966090,4.404016,26.368343,5.229238e+06
min,0.000000,19.000000,161.000000,3.088800e+04
25%,5.000000,24.000000,200.000000,1.044792e+06
50%,13.000000,26.000000,220.000000,2.839073e+06
75%,25.000000,30.000000,240.000000,6.500000e+06
max,99.000000,40.000000,307.000000,2.500000e+07


#### Task 4: Return the first 3 rows of the dataframe
 - make use of `iloc()` and `head()` functions

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,6-6,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,NaN


,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,6-6,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,NaN


#### Task 5: Select just the `Name` and `College` columns from the dataframe

,Name,College
0,Avery Bradley,Texas
1,Jae Crowder,Marquette
2,John Holland,Boston University
3,R.J. Hunter,Georgia State
4,Jonas Jerebko,NaN


,Name,College
0,Avery Bradley,Texas
1,Jae Crowder,Marquette
2,John Holland,Boston University
3,R.J. Hunter,Georgia State
4,Jonas Jerebko,NaN


#### Task 6: Select the data in rows `[3, 4, 8]` and in columns `['Position', 'Height']`

,Position,Height
3,SG,6-5
4,PF,6-10
8,PG,6-2


#### Task 7: Select only the rows where `Age`  is greater than 37.

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
101,Paul Pierce,Los Angeles Clippers,34.0,SF,38.0,6-7,235.0,Kansas,3376000.0
102,Pablo Prigioni,Los Angeles Clippers,9.0,PG,39.0,6-3,185.0,NaN,947726.0
256,Jason Terry,Houston Rockets,31.0,SG,38.0,6-2,185.0,Arizona,947276.0
261,Vince Carter,Memphis Grizzlies,15.0,SG,39.0,6-6,220.0,North Carolina,4088019.0
298,Tim Duncan,San Antonio Spurs,21.0,C,40.0,6-11,250.0,Wake Forest,5250000.0
299,Manu Ginobili,San Antonio Spurs,20.0,SG,38.0,6-6,205.0,NaN,2814000.0
304,Andre Miller,San Antonio Spurs,24.0,PG,40.0,6-3,200.0,Utah,250750.0
400,Kevin Garnett,Minnesota Timberwolves,21.0,PF,40.0,6-11,240.0,NaN,8500000.0
420,Nazr Mohammed,Oklahoma City Thunder,13.0,C,38.0,6-10,250.0,Kentucky,222888.0


#### Task 8:  Select the rows where the `Team` is `San Antonio Spurs` and the `Age` is less than 30.

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
295,Kyle Anderson,San Antonio Spurs,1.0,SF,22.0,6-9,230.0,UCLA,1142880.0
300,Danny Green,San Antonio Spurs,14.0,SG,28.0,6-6,215.0,North Carolina,10000000.0
301,Kawhi Leonard,San Antonio Spurs,2.0,SF,24.0,6-7,230.0,San Diego State,16407500.0
302,Boban Marjanovic,San Antonio Spurs,40.0,C,27.0,7-3,290.0,NaN,1200000.0
305,Patty Mills,San Antonio Spurs,8.0,PG,27.0,6-0,185.0,Saint Mary's,3578947.0
307,Jonathon Simmons,San Antonio Spurs,17.0,SG,26.0,6-6,195.0,Houston,525093.0


#### Task 9: Select the rows where the `Weight` is between 245 and 250

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
27,Henry Sims,Brooklyn Nets,14.0,C,26.0,6-10,248.0,Georgetown,947276.0
39,Kyle O'Quinn,New York Knicks,9.0,PF,26.0,6-10,250.0,Norfolk State,3750000.0
49,Joel Embiid,Philadelphia 76ers,21.0,C,22.0,7-0,250.0,Kansas,4626960.0
51,Richaun Holmes,Philadelphia 76ers,22.0,PF,22.0,6-10,245.0,Bowling Green,1074169.0
52,Carl Landry,Philadelphia 76ers,7.0,PF,32.0,6-9,248.0,Purdue,6500000.0
61,Bismack Biyombo,Toronto Raptors,8.0,C,23.0,6-9,245.0,NaN,2814000.0
65,James Johnson,Toronto Raptors,3.0,PF,29.0,6-9,250.0,Wake Forest,2500000.0
73,Jason Thompson,Toronto Raptors,1.0,PF,29.0,6-11,250.0,Rider,245177.0
91,Cole Aldrich,Los Angeles Clippers,45.0,C,27.0,6-11,250.0,Kansas,1100602.0
92,Jeff Ayres,Los Angeles Clippers,19.0,PF,29.0,6-9,250.0,Arizona State,111444.0


,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,6-6,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,NaN
3,R.J. Hunter,Boston Celtics,28.0,SG,22.0,6-5,185.0,Georgia State,1148640.0
4,Jonas Jerebko,Boston Celtics,8.0,PF,29.0,6-10,231.0,NaN,5000000.0
5,Amir Johnson,Boston Celtics,90.0,PF,29.0,6-9,240.0,NaN,12000000.0
6,Jordan Mickey,Boston Celtics,55.0,PF,21.0,6-8,235.0,LSU,1170960.0
7,Kelly Olynyk,Boston Celtics,41.0,C,25.0,7-0,238.0,Gonzaga,2165160.0
8,Terry Rozier,Boston Celtics,12.0,PG,22.0,6-2,190.0,Louisville,1824360.0
9,Marcus Smart,Boston Celtics,36.0,PG,22.0,6-4,220.0,Oklahoma State,3431040.0


#### Task 10: Change the `Height` in row 2 from 6-6 to 7-7

In [87]:
# now check the change

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,7-7,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,NaN
3,R.J. Hunter,Boston Celtics,28.0,SG,22.0,6-5,185.0,Georgia State,1148640.0
4,Jonas Jerebko,Boston Celtics,8.0,PF,29.0,6-10,231.0,NaN,5000000.0


#### Task 11: Display a count of NaNs for each column

Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary      12
dtype: int64

#### Task 12:  Fill the missing salary values with mean

Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary       0
dtype: int64

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.00,PG,25.00,6-2,180.00,Texas,7730337.00
1,Jae Crowder,Boston Celtics,99.00,SF,25.00,7-7,235.00,Marquette,6796117.00
2,John Holland,Boston Celtics,30.00,SG,27.00,6-5,205.00,Boston University,4842684.11
3,R.J. Hunter,Boston Celtics,28.00,SG,22.00,6-5,185.00,Georgia State,1148640.00
4,Jonas Jerebko,Boston Celtics,8.00,PF,29.00,6-10,231.00,NaN,5000000.00


#### Task 13: Calculate the sum of salary  column

2217949320.264574

#### Task 14: Calculate the mean `Age` 

27

#### Task 15: Sort df by the values in the `Salary` in decending order
- view the first 20 records

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
109,Kobe Bryant,Los Angeles Lakers,24.00,SF,37.00,6-6,212.00,NaN,25000000.00
169,LeBron James,Cleveland Cavaliers,23.00,SF,31.00,6-8,250.00,NaN,22970500.00
33,Carmelo Anthony,New York Knicks,7.00,SF,32.00,6-8,240.00,Syracuse,22875000.00
251,Dwight Howard,Houston Rockets,12.00,C,30.00,6-11,265.00,NaN,22359364.00
339,Chris Bosh,Miami Heat,1.00,PF,32.00,6-11,235.00,Georgia Tech,22192730.00
100,Chris Paul,Los Angeles Clippers,3.00,PG,31.00,6-0,175.00,Wake Forest,21468695.00
414,Kevin Durant,Oklahoma City Thunder,35.00,SF,27.00,6-9,240.00,Texas,20158622.00
164,Derrick Rose,Chicago Bulls,1.00,PG,27.00,6-3,190.00,Memphis,20093064.00
349,Dwyane Wade,Miami Heat,3.00,SG,34.00,6-4,220.00,Marquette,20000000.00
98,DeAndre Jordan,Los Angeles Clippers,6.00,C,27.00,6-11,265.00,Texas A&M,19689000.00


,Name,Team,Number,Position,Age,Height,Weight,College,Salary
1,Jae Crowder,Boston Celtics,99.00,SF,25.00,7-7,235.00,Marquette,6796117.00
455,Tibor Pleiss,Utah Jazz,21.00,C,26.00,7-3,256.00,NaN,2900000.00
302,Boban Marjanovic,San Antonio Spurs,40.00,C,27.00,7-3,290.00,NaN,1200000.00
40,Kristaps Porzingis,New York Knicks,6.00,PF,20.00,7-3,240.00,NaN,4131720.00
322,Walter Tavares,Atlanta Hawks,22.00,C,24.00,7-3,260.00,NaN,1000000.00
275,Alexis Ajinca,New Orleans Pelicans,42.00,C,28.00,7-2,248.00,NaN,4389607.00
235,Salah Mejri,Dallas Mavericks,50.00,C,29.00,7-2,245.00,NaN,525093.00
111,Roy Hibbert,Los Angeles Lakers,17.00,C,29.00,7-2,270.00,Georgetown,15592217.00
329,Spencer Hawes,Charlotte Hornets,0.00,PF,28.00,7-1,245.00,Washington,6110034.00
447,Rudy Gobert,Utah Jazz,27.00,C,23.00,7-1,245.00,NaN,1175880.00
